README
mod_dav_acl
-----------
-an RFC3744 implementation based on Apache (tested in linux only)

-requires patches to Apache, mostly to mod_dav module, patch(es) included.
Patch contains strong ETag handling, usec timestamps to ETags (libapr-patch,
latest releases support already usec timestamps), conditional PUTs (fixed after
2.2.8 apache) and some hooks to allow integration of ACL handlings. Apply the
patch to Apache sources and compile it before using this module.

-apr-util patch is required with the 2.2.12 -> series of apache patches

-with the old apr library the apr patch enables nano (actually u) second
timestamps in resource handlings. Patch is optional (etags aren't just that
strong without it), and you can add both inode and size directives to etags to
make them stronger.

-mod_dav_acl must be loaded after mod_dav since it uses symbols from mod_dav

-optionally (compile time option, build tests xattr existance) ACL (meta)info is
stored within the extended attributes of the filesystem. Currently XFS probably
the most convenient to work with (65k EA limit + nanosecond timestamps enabled
by default). In ext3/4 filesystem EAs can be enabled with the option user_xattr.
It is preferable to utilize EAs since they are faster than using the "standard"
property database (especially writing seems to be much faster). This is relevant
e.g. with basic puts, since the owner of a resource is stored allways although
acls were not otherwise utilized at all (== acl-method not used in practice). If
EA usage fails or aren't supported, standard db is used.

-principals are stored into a shared memory segment to allow faster ACL
processing, though some other optimizations would be nice as well...

-principal uris are either of type: <http://example.com/principals/[user]> or
<http://example.com/principals/[user]/self>. The latter format allows users
to create their own groups.


Configuration directives:
-------------------------
AclLockFile  /var/tmp/davacl.lock
-lock filename for shared memory

AclSharedMemFile /var/tmp/davacl.shm
-shared memory name

AclSharedMemSize 65000
-max size of shared memory segment, contains the list of principals (uris)

AclPrincipals http://example.com/principals
-base URI for principals

AclPrincipalDir /var/principals
-real absolute local directory path for principals

AclOwnerFullRights on/off
-the owner (if exists) has full rights to the resource if not any acl is set.

DAVETagResponse on/off
-enable strong ETag responses (required e.g. by CalDAV/XCAP/CardDAV)

DAVACL on/off
-enable ACL handling

AclUsePropertyDB on/off
-store ACL info into standard property database instead of extended attributes
of a filesysten (default on without xattr enabled builds). off == EAs activated.

AclAggregatedCurrentUserPrivilegeSet on/off
-"current-user-privilege-set" property aggregated into read privilege

AclAggregatedAcl on/off
-"acl" property aggregated into read privilege.

in debian:"sudo a2enmod dav_acl" enables module loading and basic
configuration


Principal editings:
-------------------
-a simple shell script: acl-principals included for adding/removing principals

-based on sending simple http (dav) requests to the server. Test client utilizes
libcurl library.

-does not include adding user identities into the password database (htdigest),
so users (+credentials) must be added separetely.


Apache patching (must still be done, patch integration going on):
-----------------------------------------------------------------
-with 2.2.12 apache or never, patch apr-util package first

-apply one of the included patches to your Apache sources, e.g.:
cd to/apache/src/directory
patch -p1 < httpd-2.2.xx-ju.patch  (ju patch contains all httpd stuff)
with <= 2.2.16 (separated patch set)
patch -p1 < httpd-2.2.12-dav.patch
patch -p1 < httpd-2.2.12-etag.patch
patch -p1 < httpd-2.2.12-liveprops.patch
patch -p1 < httpd-2.2.12-repos.patch
if you compile from the trunk repository, liveprops and repos have been applied,
so there's no need to try to apply those patches.
-build with with dpkg-buildpackage/rpm-build (preferred) or by "standard"
configure && make && make install
-install patched apache including the development package, preferably using
distributions packaging tools (dpkg or rpm)

Install mod_dav_acl:
--------------------
-compile with dpkg-buildpackage or rpm-build (preferred) or by "standard"
./configure && make
-install by using dpkg (-i ..), rpm (-Uvh ...) or by make install
-you might also need to regenerate the ./configure script with autogen.sh.


Configurations:
---------------
-with debian/ubuntu you can enable dav_acl with a2enmod, with rpm based distros
you probably need some manual editings for loadings of modules.
-configure the usage of dav_acl within appropriate directories, and create
required users (htdigest) and principals (acl-principals), and make sure apache
daemon has appropriate access rights to those directories.


Installation sequence in debian-based distro:
---------------------------------------------
   apt-get source apr-util
   cd to-src-directory
   patch apr-util && update debian/changelog release name (to simplify seeing
      patched versions of installed packages)
   build: dpkg-buildpackage -rfakeroot -uc -us -b
   install bin & dev packages: sudo dpkg -i ...

   apt-get source apache2
   cd to-src-directory && patch apache && update debian/changelog
   build: dpkg-buildpackage -rfakeroot -uc -us -b
   install required bin, lib, dev packages: sudo dpkg -i ...

   cd to-mod-dav-acl-directory
   build: dpkg-buildpackage -rfakeroot -uc -us -b
   install module (and dev) packages: sudo dpkg -i ...

   configure apache, install credentials for users/testers and enable required
   modules, and run tests e.g. with make check


Tests:
------
-once the module is successfully compiled and loaded, tests can be run with
make check or by running ./acl_tests. A simple libcurl testclient will send some
basic tests to a server listening at localhost:8080.

-you need to create (digest) users: joe, joppe and admin before running the
tests (all share the same password == password) unless you modify the scripts,
of course. Also /xxx/principals and /xxx/acltests directories must be created
and accessable by apache/httpd (in debian-based distros 'www-data' user, and
with fedora 'apache').

Example dav_acl.conf for testing purposes:

# for local testing only
Listen 127.0.0.1:8080

<VirtualHost 127.0.0.1:8080>
    DocumentRoot /var/www/testing/www
    ServerName localhost

Alias /principals "/var/www/testing/principals"
Alias /acltests "/var/www/testing/acltests/"

AclOwnerFullRights on
AclLockFile        /var/tmp/davacl.lock
AclSharedMemFile   /var/tmp/davacl.shm
AclSharedMemSize   65000
AclPrincipals      http://localhost:8080/principals
AclPrincipalDir    /var/www/testing/principals
AclUsePropertyDB   off
AclAggregatedCurrentUserPrivilegeSet on
AclAggregatedAcl   off

<Location /principals>
   Dav on

   AuthType Digest
   AuthName "testing"
   AuthUserFile /var/www/passwords
   Require valid-user
   AuthDigestProvider file

   FileETag MTime

   DAVETagResponse on
   DAVACL on

</Location>

<Location /acltests>
   Dav on

   AuthType Digest
   AuthName "testing"
   AuthUserFile /var/www/passwords
   Require valid-user
   AuthDigestProvider file

   FileETag MTime

   DAVETagResponse on
   DAVACL on

</Location>

</VirtualHost>



ChangeLog

Changes with 0.3.0

  *) Correct parameters to dav_new_error() for httpd v2.4.
     [Graham Leggett <minfrin@sharp.fm>]


look debian/changelog about later updates

02.07.2008 Jari Urpalainen (0.1.4)
	* httpd-2.2.8-ju.patch:
	-fixed struct misuse (Thanks to Romain Tartière)
	-cleaned up 2.2.8 changes, i.e. ETag handlings 
	* acl_tests & acl_files:
	-added some new tests 
	* README, dac_acl.conf:
	-updated docs and initial configuration

05.06.2008 Jari Urpalainen (0.1.3)
	* mod_dav_acl.c:
	-shared memory usage of the first principal bug fixed (multiple processes got out-of-sync)
	-in principal property search wrong order of strstr pointers
	-added one test for principal property search
	-apache patch for the version 2.2.8 created

19.02.2008 Jari Urpalainen
	* mod_dav_acl.c:
	-fixed body reading with multiple threads
	-fixed Depth header handling bug
	
25.10.2007 Jari Urpalainen
	* debian/*
	-added debian packaging
	* mod_dav_acl.c:
	-fixed an empty <displayname> bug
	* apache-patch/mod_dav.c:
	-in ETag response, removed generation of vlist_validator

11.10.2007 Jari Urpalainen
	-COPYING content SHOULD be LGPL text
	* principals.sh
	-using test_cli for sending principal creations etc.
	* configure.in:
	-more debian friendly, apxs2 testing & libxattr-devel (thanks to Aki
	Niemi)  

02.07.2007 Jari Urpalainen
	* mod_dav_acl.c:
	-fixed memory leaks when using filters and all principals resources 
	were not freed when resources were re-read
	-parent file resources are read with a new public api get_pathname 
	* acl_liveprops.c:
	-fixed a bug which reported wrong acl properties
	* apache/modules/mod_dav.h:
	-added public api:get_pathname 
	* test-cli.c: 
	-added new test client instead of plain curl as command line
	-parses e.g. ETag values from responses
	* acl_files:
	-changed tests accordingly based on test_cli client

26.03.2007 Jari Urpalainen
	* Initial release 0.1.0


acl_private.h
/**
 * This is part of a mod_dav_acl library.
 *
 * Copyright (C) 2006 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef _DAV_PRIVATE_H_
#define _DAV_PRIVATE_H_

#define NS_DAV  "DAV:"

#define DBHOOKS(conf)  (conf && conf->provider) ? conf->provider->propdb : NULL

#define ARRAY_SIZE(a) sizeof(a) / sizeof(a[0])
#define ARRAY(a) a,ARRAY_SIZE(a)

#define FOR_CHILD(node, parent) \
	for (node = parent ? parent->children : NULL; node; node = node->next)

#define NODE_NS(node) (node->ns && node->ns->href && \
			strcmp((char *) node->ns->href, NS_DAV) == 0)

#define NODE_NOT_DAV(node) node->type != XML_ELEMENT_NODE || !NODE_NS(node)

#define NODE_MATCH(node, y) (strcmp((char *) node->name, y) == 0)

/* directory configuration data */
typedef struct _davacl_dir_cfg {
    char *shared_mem_file;		/* shared mem file name */
    char *principals;			/* http://example.com/principals */
    char *principal_dir;		/* local directory for principals */

    int shared_mem;			/* shared memory max size for
					 * principals */
    unsigned owner_full_rights:1,	/* owner has full rights if not any
					 * acl is set */
	     use_std_property_db:1,	/* store acl as standard dead
					 * property */
	     acl_aggregated:1,		/* acl aggregated into read */
	     cups_aggregated:1;		/* current-user-privilege-set aggregated
					 * into read privilege */

    apr_shm_t *shm;			/* shared memory segment */
    apr_rmm_t *rmm;			/* rmm pool within shm */

    apr_rmm_off_t *off_user;		/* first user record offset in the
					 * shared memory area */
    apr_pool_t *pool;			/* pool */

    const dav_provider *provider;	/* filesystem provider */

} davacl_dir_cfg;

/* server configuration data */
typedef struct _davacl_server_cfg {
    char *lock_file;			/* mutex lock file name */
    apr_global_mutex_t *mutex;		/* global mutex for shared memory */
} davacl_server_cfg;

int acl_get_acl(const dav_resource *resource, const davacl_dir_cfg *conf,
                const char *filename, const char **ppb, int *c);

const char *acl_get_principal_uri(request_rec *r, const davacl_dir_cfg *conf);
const char *acl_get_owner(const dav_resource *resource,
                          const davacl_dir_cfg *conf, const char *filename);

const char *acl_get_group(const dav_resource *resource,
                          const davacl_dir_cfg *conf);


static inline void acl_lock(davacl_server_cfg *sconf)
{
    if (sconf && sconf->mutex)
	apr_global_mutex_lock(sconf->mutex);
}

static inline void acl_unlock(davacl_server_cfg *sconf)
{
    if (sconf && sconf->mutex)
	apr_global_mutex_unlock(sconf->mutex);
}


#define ADDRESS(conf, off)		apr_rmm_addr_get(conf->rmm, off)
#define UID(conf, off)			(acl_uid_t *) (ADDRESS(conf, off))
#define USERNAME(conf, off)		(char *) (ADDRESS(conf, off))
#define GET_USERNAME(conf, off)		off ? USERNAME(conf, off) : NULL

/* principal object structure in shared memory */
typedef struct
{
    apr_rmm_off_t next,		/* next struct pointer */
		  user,		/* principal uri */
		  displayname,	/* displayname */
		  children;	/* group member list */
} acl_uid_t;

#define LOOP_MAX 20

#if DEBUG
    #define TRACE(r, arg...) \
	ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, r->server, ##arg)
    #define TRACE_WARNING(r, arg...) \
	ap_log_error(APLOG_MARK, APLOG_WARNING, 0, r->server, ##arg)
    #define TRACE_ERROR(r, arg...) \
	ap_log_error(APLOG_MARK, APLOG_CRIT, 0, r->server, ##arg)
#else
     #define TRACE(r, arg...) do {} while (0)
     #define TRACE_WARNING(r, arg...) do {} while (0)
     #define TRACE_ERROR(r, arg...) do {} while (0)
#endif

dav_error *acl_check_req(request_rec *r, const char *filename,
                         const dav_resource *resource,
                         const davacl_dir_cfg *conf,
                         const dav_prop_name *name, int count, int count_loop);

const char *acl_get_principal_dir(const davacl_dir_cfg *conf);

int acl_is_user_group_member(const char *user, const char *group,
                             const davacl_dir_cfg *conf);

void acl_store_owner(request_rec *r, const dav_resource *resource,
                     davacl_dir_cfg *conf);

void acl_update_principal(request_rec *r, const dav_resource *resource,
                          davacl_dir_cfg *conf);

void acl_update_all_principals(request_rec *r, davacl_dir_cfg *conf);

void acl_shm_init(const request_rec *r, davacl_dir_cfg *conf);

dav_error *acl_store_acl(request_rec *r, dav_resource *resource,
                         davacl_dir_cfg *conf, dav_buffer *buffer);

#define REPOS(conf) (conf && conf->provider) ? conf->provider->repos : NULL

#define MAX_RECURSION 10

#endif /* _DAV_PRIVATE_H_ */



mod_dav_acl.h
/**
 * This is part of a mod_dav_acl library.
 *
 * Copyright (C) 2006 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef _MOD_DAV_ACL_H_
#define _MOD_DAV_ACL_H_

#define XML_VERSION "1.0"
#define DAV_ACL		"user.davacl"
#define DAV_OWNER	"user.owner"

#define NS_DAV		"DAV:"

#define ARRAY_SIZE(a) sizeof(a) / sizeof(a[0])
#define ARRAY(a) a,ARRAY_SIZE(a)

#define DAV_DEFAULT_PROVIDER "filesystem"

#ifdef __cplusplus
extern "C" {
#endif

DAV_DECLARE(const char *)	dav_acl_get_auth_principal(const dav_resource *resource);
DAV_DECLARE(int)		dav_acl_get_acl(const dav_resource *resource, const char **ppb, int *c);
DAV_DECLARE(void)		dav_acl_get_aggregated(const dav_resource *resource, int *acl, int *cups);

DAV_DECLARE(const char *)	dav_acl_get_principal_dir(request_rec *r);
DAV_DECLARE(const char *)	dav_acl_get_principals(request_rec *r);
DAV_DECLARE(const char *)	dav_acl_get_privs(const dav_resource *resource);
DAV_DECLARE(int)		dav_acl_is_resource_principal(const dav_resource *resource);
DAV_DECLARE(const char *)	dav_acl_get_group_membership(const dav_resource *resource);
DAV_DECLARE(const char *)	dav_acl_get_owner(const dav_resource *resource);

DAV_DECLARE(dav_error*)		dav_acl_store_owner(request_rec *r, const dav_resource *resource);

DAV_DECLARE(dav_error*)		dav_acl_check(request_rec *r, const dav_resource *resource,
					      const dav_prop_name *p, int c);

DAV_DECLARE(apr_xml_doc *)	dav_acl_get_prop_doc(request_rec *r, const xmlNode *n);

/** input filter */
typedef struct
{
    apr_bucket_brigade *pbb;
    dav_buffer buffer;
    request_rec *r;
} dav_acl_input_filter_t;

DAV_DECLARE(apr_status_t)	dav_acl_input_filter(ap_filter_t *f, apr_bucket_brigade *pbb_out, ap_input_mode_t emode, apr_read_type_e eblock, apr_off_t nbytes);

DAV_DECLARE(int)		dav_acl_read_body(request_rec *r, dav_buffer *buffer);

DAV_DECLARE(const char *)	dav_acl_get_prop(request_rec *r, const dav_resource *resource, const dav_provider *provider, const dav_prop_name *prop);

DAV_DECLARE(dav_error *)	dav_acl_privilege_error(request_rec *r, const char *pch, const char *desc, ...)
	__attribute__((__format__ (__printf__, 3, 4)));

DAV_DECLARE(int)		dav_acl_exec_error(request_rec *r, dav_error *err);

DAV_DECLARE(void)		dav_acl_last_mtime(const char *subdir, request_rec *r, apr_pool_t *pool, int recur);

#ifdef __cplusplus
}
#endif

#endif /* _MOD_DAV_ACL_H_ */



acl_liveprops.h
/**
 * This is part of a mod_dav_acl library.
 * 
 * Copyright (C) 2006 Nokia Corporation.
 * 
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */
#ifndef _ACL_LIVEPROPS_H_
#define _ACL_LIVEPROPS_H_

void dav_acl_gather_propsets(apr_array_header_t *uris);
int dav_acl_find_liveprop(const dav_resource *resource, const char *ns_uri,
                          const char *name, const dav_hooks_liveprop **hooks);
void dav_acl_insert_all_liveprops(request_rec *r, const dav_resource *resource,
                                  dav_prop_insert what, apr_text_header *phdr);

/* register our live property URIs with mod_dav. */
void dav_acl_register_props(apr_pool_t *p);

#endif /* _ACL_LIVEPROPS_H_ */


acl_liveprops.c
/**
 * This is part of a mod_dav_acl library.
 * acl_liveprops.c: mod_dav_acl live property provider functions
 *
 * Copyright (C) 2006 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <sys/types.h>

#include <httpd.h>
#include <libxml/tree.h>

#include <mod_dav.h>

#include "mod_dav_acl.h"
#include "apr_strings.h"

/*
** The namespace URIs that we use. This list and the enumeration must
** stay in sync.
*/
static const char * const dav_acl_namespace_uris[] =
{
    "DAV:",
    NULL	/* sentinel */
};

enum {
    DAV_ACL_NAMESPACE_URI_DAV = 0  /* the DAV: namespace URI */
};

#define ACL_RO_PROP(name, enum_name) \
	{ DAV_ACL_NAMESPACE_URI_DAV, name, ACL_PROPID_##enum_name, 0 }
#define ACL_RW_PROP(name, enum_name) \
	{ DAV_ACL_NAMESPACE_URI_DAV, name, ACL_PROPID_##enum_name, 1 }

enum {
    ACL_PROPID_supported_privilege_set = 1,
    ACL_PROPID_current_user_privilege_set,
    ACL_PROPID_acl,
    ACL_PROPID_acl_restrictions,
    ACL_PROPID_inherited_acl_set,
    ACL_PROPID_principal_collection_set,
    ACL_PROPID_owner,
    ACL_PROPID_principal_url,
    ACL_PROPID_alternate_uri_set,
    ACL_PROPID_group_membership,
    ACL_PROPID_group_member_set,
    ACL_PROPID_group,
    ACL_PROPID_current_user_principal
};

static const dav_liveprop_spec dav_acl_props[] =
{
    ACL_RO_PROP("owner", owner),   /* could also be a writable resource,
				    * now set by put or mkcol */
    ACL_RW_PROP("group", group),
    ACL_RO_PROP("supported-privilege-set", supported_privilege_set),
    ACL_RO_PROP("current-user-privilege-set", current_user_privilege_set),
    ACL_RO_PROP("acl", acl),
    ACL_RO_PROP("acl-restrictions", acl_restrictions),
    ACL_RO_PROP("inherited-acl-set", inherited_acl_set),
    ACL_RO_PROP("principal-collection-set", principal_collection_set),
    ACL_RO_PROP("principal-URL", principal_url),
    ACL_RO_PROP("alternate-URI-set", alternate_uri_set),
    ACL_RO_PROP("group-membership", group_membership),
    ACL_RW_PROP("group-member-set", group_member_set),
    ACL_RO_PROP("current-user-principal", current_user_principal),
    { 0 } /* sentinel */
};

const dav_hooks_liveprop dav_acl_hooks_liveprop;

static const dav_liveprop_group dav_acl_liveprop_group =
{
    dav_acl_props,
    dav_acl_namespace_uris,
    &dav_acl_hooks_liveprop
};

static xmlNodePtr add_supported(xmlNodePtr parent, xmlNsPtr ns,
                                const char *priv, int abstract,
                                const char *desc)
{
    xmlNodePtr ret = xmlNewChild(parent, ns,
				 (const xmlChar *) "supported-privilege", NULL);
    xmlNodePtr cur = xmlNewChild(ret, ns, (const xmlChar *) "privilege", NULL);
    xmlNewChild(cur, ns, (const xmlChar *) priv, NULL);
    if (abstract)
	xmlNewChild(ret, ns, (const xmlChar *) "abstract", NULL);
    xmlNewChild(ret, ns, (const xmlChar *) "description",
		(const xmlChar *) desc);
    return ret;
}

static dav_prop_insert dav_acl_insert_prop(const dav_resource *resource,
                                           int propid, dav_prop_insert what,
                                           apr_text_header *phdr)
{
    const char *value = NULL;
    const char *s = NULL;
    apr_pool_t *p = resource->pool;
    const dav_liveprop_spec *info;
    int global_ns, rc;

    if (!resource->exists)
	return DAV_PROP_INSERT_NOTDEF;

    /* ### we may want to respond to DAV_PROPID_resourcetype for PRIVATE
       ### resources. need to think on "proper" interaction with mod_dav */

    switch (propid) {
    case ACL_PROPID_acl:
	if (what == DAV_PROP_INSERT_VALUE) {
	    /* request_rec *r = resource->hooks->get_request_rec(resource) */;
	    xmlBufferPtr buf = NULL;
	    const char *pch = NULL;
	    xmlDocPtr doc;
	    xmlNodePtr node;

	    rc = dav_acl_get_acl(resource, &pch, &rc);
	    if (rc < 0)
		return DAV_PROP_INSERT_NOTDEF;

	    doc = xmlParseMemory(pch, rc);
	    node = doc && doc->children ? doc->children : NULL;

	    buf = xmlBufferCreate();
	    xmlNodeDump(buf, doc, node, 0, 1);
	    xmlFreeDoc(doc);

	    apr_text_append(p, phdr, apr_psprintf (p, "%s" DEBUG_CR,
			    buf->content));
	    /* we inserted whatever was asked for */
	    xmlBufferFree(buf);
	    return what;
	}
	break;

    case ACL_PROPID_acl_restrictions:
	value = "<D:required-principal>"
		"<D:authenticated/>"
		"<D:unauthenticated/>"
		"<D:self/>"
		"<D:all/>"
		"<D:href/>"
		"<D:property><D:owner/></D:property>"
		"<D:property><D:group/></D:property>"
		"</D:required-principal>";
	break;

    case ACL_PROPID_supported_privilege_set:
	if (what == DAV_PROP_INSERT_VALUE) {
	    xmlBufferPtr buf = NULL;

	    xmlDocPtr doc = xmlNewDoc((const xmlChar *) XML_VERSION);
	    xmlNsPtr ns;
	    xmlNodePtr cur, child, node;
	    int acl, cups;

	    dav_acl_get_aggregated(resource, &acl, &cups);

	    doc->children = xmlNewDocNode(doc, NULL,
					  (const xmlChar *) "root", NULL);
	    xmlSetNs(doc->children,
		     ns = xmlNewNs(doc->children, (const xmlChar *) "DAV:",
		     (const xmlChar *) "D"));

	    node = xmlNewChild(doc->children, ns,
				(const xmlChar *) "supported-privilege-set",
				NULL);

	    cur = add_supported(node, ns, "all", 0, "All privileges");

	    child = add_supported(cur, ns, "read", 0, "Read");

	    add_supported(acl ? child : cur, ns, "read-acl", 0, "Read ACL");

	    add_supported(cups ? child : cur, ns,
			  "read-current-user-privilege-set", 0,
			  "Read Current User");

	    child = add_supported(cur, ns, "write", 0, "Write");

	    add_supported(child, ns, "write-acl", 0, "Write ACL");
	    add_supported(child, ns, "write-content", 0, "Write content");
	    add_supported(child, ns, "write-properties", 0,
							"Write properties");

	    if (resource->collection) {
		add_supported(child, ns, "bind", 0, "Create a collection");
		add_supported(child, ns, "unbind", 0, "Remove a collection");
	    }
	    add_supported(cur, ns, "unlock", 0, "Unlock");

	    buf = xmlBufferCreate();
	    xmlNodeDump(buf, doc, node, 0, 1);
	    xmlFreeDoc(doc);

	    apr_text_append(p, phdr, apr_psprintf(p, "%s" DEBUG_CR,
							buf->content));
	    /* we inserted whatever was asked for */
	    xmlBufferFree(buf);
	    return what;
	}
	break;

    case ACL_PROPID_group_membership:
	value = dav_acl_get_group_membership(resource);
	if (value == NULL)
	    return DAV_PROP_INSERT_NOTDEF;
	break;

    case ACL_PROPID_owner:
	value = dav_acl_get_owner(resource);
	if (value == NULL)
	    return DAV_PROP_INSERT_NOTDEF;
	break;

    case ACL_PROPID_alternate_uri_set:
	value = "";  /* may be empty */
	break;

    case ACL_PROPID_inherited_acl_set:
	value = "";  /* not supported because of complexity... */
	break;

    case ACL_PROPID_current_user_privilege_set:
	value = dav_acl_get_privs(resource);
	break;

    case ACL_PROPID_principal_collection_set:
	{
	    request_rec *r = resource->hooks->get_request_rec(resource);

	    const char *pcsz = dav_acl_get_principals(r);

	    value = apr_psprintf(p, "<D:href>%s/</D:href>", pcsz ? pcsz : "");
	}
	break;

    case ACL_PROPID_current_user_principal:
	value = dav_acl_get_auth_principal(resource);
	if (value)
	    value = apr_psprintf(p, "<D:href>%s</D:href>", value);
	else
	    value = "<D:unauthenticated/>";
	break;

    default:
	/* ### what the heck was this property? */
	return DAV_PROP_INSERT_NOTDEF;
    }

    /* assert: value != NULL */

    /* get the information and global NS index for the property */
    global_ns = dav_get_liveprop_info(propid, &dav_acl_liveprop_group, &info);

    /* assert: info != NULL && info->name != NULL */

    if (what == DAV_PROP_INSERT_VALUE)
	s = apr_psprintf(p, "<lp%d:%s>%s</lp%d:%s>" DEBUG_CR,
			 global_ns, info->name, value, global_ns, info->name);
    else if (what == DAV_PROP_INSERT_NAME)
	s = apr_psprintf(p, "<lp%d:%s/>" DEBUG_CR, global_ns, info->name);
    else
	/* assert: what == DAV_PROP_INSERT_SUPPORTED */
	s = apr_psprintf(p,
			 "<D:supported-live-property D:name=\"%s\" "
			 "D:namespace=\"%s\"/>" DEBUG_CR,
			 info->name, dav_acl_namespace_uris[info->ns]);

    apr_text_append(p, phdr, s);

    /* we inserted whatever was asked for */
    return what;
}

static int dav_acl_is_writable(const dav_resource *resource, int propid)
{
    const dav_liveprop_spec *info;

    dav_get_liveprop_info(propid, &dav_acl_liveprop_group, &info);

    return info->is_writable;
}

static char *acl_get_elem_text(apr_pool_t *p, const apr_text *t)
{
    char *pch = NULL;

    for ( ; t; t = t->next)
	pch = apr_pstrcat(p, pch ? pch : "", t->text, NULL);

    return pch;
}

static dav_error *dav_acl_patch_validate(const dav_resource *resource,
                                         const apr_xml_elem *elem,
                                         int operation, void **context,
                                         int *defer_to_dead)
{
    dav_elem_private *priv = elem->priv;

    switch (priv->propid) {
    case ACL_PROPID_group:
    case ACL_PROPID_group_member_set:
	/* actually a dead property, but defined as alive to have
	 * this callback for proppatch */
	*defer_to_dead = TRUE;

	if (operation == DAV_PROP_OP_SET) {
	    request_rec *r = resource->hooks->get_request_rec(resource);
	    const char *pch = dav_acl_get_principals(r);
	    apr_xml_elem *child = elem->first_child;

	    for ( ; child; child = child->next) {
		if (strcmp(child->name, "href") == 0) {
		    const char *
			href = acl_get_elem_text(resource->pool,
						 child->first_cdata.first);

		    if (pch && (!href || strncmp(href, pch, strlen (pch)) != 0))
			return dav_new_error(resource->pool, HTTP_CONFLICT, 0,
					     "The principal uri path not "
					     "supported");
		}
	    }
	}
	break;

    default:
	break;
    }

    return NULL;
}

static dav_error *dav_acl_patch_exec(const dav_resource *resource,
                                     const apr_xml_elem *elem,
                                     int operation, void *context,
                                     dav_liveprop_rollback **rollback_ctx)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
    return NULL;
}

static void dav_acl_patch_commit(const dav_resource *resource, int operation,
                                 void *context,
                                 dav_liveprop_rollback *rollback_ctx)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
}

static dav_error *dav_acl_patch_rollback(const dav_resource *resource,
                                         int operation, void *context,
                                         dav_liveprop_rollback *rollback_ctx)
{
    /* NOTE: this function will not be called unless/until we have
       modifiable (writable) live properties. */
    return NULL;
}

const dav_hooks_liveprop dav_acl_hooks_liveprop = {
    dav_acl_insert_prop,
    dav_acl_is_writable,
    dav_acl_namespace_uris,
    dav_acl_patch_validate,
    dav_acl_patch_exec,
    dav_acl_patch_commit,
    dav_acl_patch_rollback,
};

void dav_acl_gather_propsets(apr_array_header_t *uris)
{
}

int dav_acl_find_liveprop(const dav_resource *resource,
                          const char *ns_uri, const char *name,
                          const dav_hooks_liveprop **hooks)
{
    /* don't try to find any liveprops if this isn't "our" resource
     * if (resource->hooks != &dav_acl_hooks_repos)
     *   return 0;
     */
    return dav_do_find_liveprop(ns_uri, name, &dav_acl_liveprop_group, hooks);
}

void dav_acl_register_props(apr_pool_t *p)
{
    /* register the namespace URIs */
    dav_register_liveprop_group(p, &dav_acl_liveprop_group);
}
